### PR TITLE
GLTFExporter WebWorker Support

### PR DESCRIPTION
**Pull request by robertlong**
**Description**

This PR makes the GLTFExporter work in a WebWorker with OffscreenCanvas support. All references to classes/methods not available on the WorkerGlobalScope have been removed or replaced.

This contribution is funded by [Matrix.org](https://matrix.org).

### CODE REVIEW

**Comment by donmccurdy** [comment]
Looks great thanks! @takahirox any concerns?

**Comment by takahirox** [comment]
I haven't tested yet but looks good to me.

**Comment by robertlong** [comment]
Let me convert this to a draft for just a bit longer. I'm still having challenges with the images being corrupted when exporting as a glb.

**Comment by robertlong** [comment]
Found the fix, forgot to return the promise correctly. Everything seems good now, ready for review 👍

**Comment by takahirox** [comment]
@robertlong Do you have online demo or test? It will help us review.

**Comment by robertlong** [comment]
Here's the example:

Live Demo: https://robertlong.github.io/three-gltf-exporter-webworker/
Repo: https://github.com/robertlong/three-gltf-exporter-webworker

File: examples/jsm/exporters/GLTFExporter.js [hunk 1]
```diff
@@ -1067,14 +1072,7 @@ class GLTFWriter {
...
 
-			if ( ( typeof HTMLImageElement !== 'undefined' && image instanceof HTMLImageElement ) ||
-				( typeof HTMLCanvasElement !== 'undefined' && image instanceof HTMLCanvasElement ) ||
-				( typeof OffscreenCanvas !== 'undefined' && image instanceof OffscreenCanvas ) ||
-				( typeof ImageBitmap !== 'undefined' && image instanceof ImageBitmap ) ) {
-
-				ctx.drawImage( image, 0, 0, canvas.width, canvas.height );
-
-			} else {
+			if ( image.data !== undefined ) {
```
Comment (donmccurdy) [inline examples/jsm/exporters/GLTFExporter.js:1075]: Is this whole block meant to handle THREE.DataTexture, do you know? Had trouble reading this (especially before), let's add a comment if that's the case.

```suggestion
			if ( image.data !== undefined ) { // THREE.DataTexture
```

File: examples/jsm/exporters/GLTFExporter.js [hunk 2]
```diff
@@ -1101,24 +1099,35 @@ class GLTFWriter {
...
+				let toBlobPromise;
+
+				if ( canvas.toBlob !== undefined ) {
+
+					toBlobPromise = new Promise( ( resolve ) => canvas.toBlob( resolve, mimeType ) );
 
-					canvas.toBlob( function ( blob ) {
+				} else {
+
+					toBlobPromise = canvas.convertToBlob( { type: mimeType } );
```
Comment (donmccurdy) [inline examples/jsm/exporters/GLTFExporter.js:1118]: That's a fun switch... 😅

**Review by donmccurdy — APPROVED** [review]

File: examples/jsm/exporters/GLTFExporter.js (hunk 1 above)
Comment (robertlong) [reply examples/jsm/exporters/GLTFExporter.js:1075]: Yeah, that's what this change is for. It looks like the [first commit with that code](https://github.com/mrdoob/three.js/commit/95e9206ab0d70e1f9b2fc496d2a9c03a56a4f831#diff-f90dfee333571797f4ddf7413faa1b10af7aa347ee7ec4bb10a5d4b87457f021R904) was meant to handle only DataTextures falling through to the `else` case. But in practice it has made it more brittle for other texture types like handling OffscreenCanvas textures in [this commit](https://github.com/mrdoob/three.js/commit/c029278aac6cec3ecb4f3ae7e09d4444a4c7664f).

I think the comment is a good idea for those not clear on what's going on. Thanks!

File: examples/jsm/exporters/GLTFExporter.js (hunk 2 above)
Comment (robertlong) [reply examples/jsm/exporters/GLTFExporter.js:1118]: Yeah, you'd think they would keep the same method name for OffscreenCanvas?

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
Minor clean up: https://github.com/mrdoob/three.js/commit/340d588ca21de2401f459e402310ec440720d686

**Comment by repalash** [comment]
Hi, was the merging of roughness and metalness map tested after this change?

It looks like the cached canvas is also being used in the `buildMetalRoughTexture` function which is set to the image source in the returned texture and is then exported by drawing that image on the same cached canvas which should result in an error.

**Comment by repalash** [comment]
Just tested on https://threejs.org/editor/ : adding a roughness map to an object and exporting gltf, saves a blank roughness/metalness texture.

**Comment by mrdoob** [comment]
Oh no...

**Comment by mrdoob** [comment]
@robertlong Any chance you can take a look?

**Comment by mrdoob** [comment]
@robertlong Never mind.

**Comment by mrdoob** [comment]
@robertlong Never mind.

**Comment by schwyzl** [comment]
Thanks for this @robertlong

**Pull request merged**